pkg/handler/handler.go

```diff
@@ -50,77 +51,79 @@ func computeETag(data []byte) string {
-	}
-	return o, nil
+	o.UpdateSpecLazy(swagger)
+
+	// Use a pass-through transformer that does nothing but record
+	// the time of last modification, since this is only called when
+	// the jsonCache is updated.
+	timeUpdate := cached.NewTransformer[*spec.Swagger](func(result cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
+		o.lastModified = time.Now()
+		return result
+	}, &o.specCache)
+	o.jsonCache = cached.NewTransformer(func(result cached.Result[*spec.Swagger]) cached.Result[[]byte] {
+		if result.Err != nil {
+			return cached.NewResultErr[[]byte](result.Err)
+		}
+		json, err := json.Marshal(result.Data)
+		if err != nil {
+			return cached.NewResultErr[[]byte](err)
+		}
+		return cached.NewResultOK(json, computeETag(json))
+	}, timeUpdate)
+	o.protoCache = cached.NewTransformer(func(result cached.Result[[]byte]) cached.Result[[]byte] {
+		if result.Err != nil {
+			return cached.NewResultErr[[]byte](result.Err)
+		}
+		proto, err := ToProtoBinary(result.Data)
+		if err != nil {
+			return cached.NewResultErr[[]byte](err)
+		}
+		// We can re-use the same etag as json because of the Vary header.
+		return cached.NewResultOK(proto, result.Etag)
+	}, o.jsonCache)
+	return o
 }
 
 func (o *OpenAPIService) getSwaggerBytes() ([]byte, string, time.Time, error) {
-	o.rwMutex.RLock()
-	defer o.rwMutex.RUnlock()
-	specBytes, err := o.jsonCache.Get()
-	if err != nil {
-		return nil, "", time.Time{}, err
-	}
-	etagBytes, err := o.etagCache.Get()
-	if err != nil {
-		return nil, "", time.Time{}, err
-	}
-	return specBytes, string(etagBytes), o.lastModified, nil
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+	result := o.jsonCache.Get()
+	return result.Data, result.Etag, o.lastModified, result.Err
 }
 
 func (o *OpenAPIService) getSwaggerPbBytes() ([]byte, string, time.Time, error) {
-	o.rwMutex.RLock()
-	defer o.rwMutex.RUnlock()
-	specPb, err := o.protoCache.Get()
-	if err != nil {
-		return nil, "", time.Time{}, err
-	}
-	etagBytes, err := o.etagCache.Get()
-	if err != nil {
-		return nil, "", time.Time{}, err
-	}
-	return specPb, string(etagBytes), o.lastModified, nil
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+	result := o.protoCache.Get()
+	return result.Data, result.Etag, o.lastModified, result.Err
 }
 
-func (o *OpenAPIService) UpdateSpec(openapiSpec *spec.Swagger) (err error) {
-	o.rwMutex.Lock()
-	defer o.rwMutex.Unlock()
-	o.jsonCache = o.jsonCache.New(func() ([]byte, error) {
-		return json.Marshal(openapiSpec)
-	})
-	o.protoCache = o.protoCache.New(func() ([]byte, error) {
-		json, err := o.jsonCache.Get()
-		if err != nil {
-			return nil, err
-		}
-		return ToProtoBinary(json)
-	})
-	o.etagCache = o.etagCache.New(func() ([]byte, error) {
-		json, err := o.jsonCache.Get()
-		if err != nil {
-			return nil, err
-		}
-		return []byte(computeETag(json)), nil
-	})
-	o.lastModified = time.Now()
-
+func (o *OpenAPIService) UpdateSpec(swagger *spec.Swagger) error {
+	o.UpdateSpecLazy(cached.NewResultOK(swagger, uuid.New().String()))
```

Comment:
That UUID here basically means "anytime someone calls this function, re-compute the json to see if it changed", i.e. invalidate the json-cache since something might have changed.

Reply:
Gotcha. I see comptueEtag used in jsonCache now:

https://github.com/kubernetes/kube-openapi/pull/371/files#diff-47683fcfa1b9a283cbc534ece58037debb4ceefd6584606550485d55c10e4dd9R88

resolved